README
sdns - Secure DNS.  

I wrote this with a few others back when I interned at the lab


Put the sdns files in /sdns - typically, like this:

	cd /;tar -xvpf sdns.tar

If you wish to use the distributed setup (you almost certainly do not) as root type:
	'sdns'

If you wish to customize sdns, read the rest of this document.

=========== Components:

README		This file

makefile	The makefile for this directory
		Type 'make' to update sdns to reflect changes

gogo		a short shell script that loops forever running sdns
		It is typically invoked from /etc/rc.d/rc.local

go		    kills the sdns process, allowing an update to be made.
		     If gogo invoked sdns, killing sdns will cause a restart.

hosts		hosts file used to create the lookup tables for sdns

defaults	default domain file that specifies default answers for whole domains

convert.pl	perl script that creates table.h from hosts and defaults

table.h		This is the part of the sdns program that specifies answers to specific
		inquiries on your system.  It is generated from 'defaults' and 'hosts'
		by convert.pl

sdns.c		program source 

sdns		generated executable 

=========== Operation:

sdns listens on port 53 for DNS requests and replies to them, logging results in
the 'logfile' file in the /sdns directory.

=========== Changes:

The easiest change to make is to copy a 'hosts' file (typically
/etc/hosts) into the /sdns directory:
	cp /etc/hosts /sdns

Then type make:
	mv sdns zz;make

Then, if the sdns service is already running via  'gogo', type:
	go

This will invoke the changes.

To change default domains, edit 'defaults' then use:
	mv sdns zz;make
	go
as above.



======================= This describes how to build dns packets with multiple resource records

-------------
ETH			taken care of by 'sendto'
-------------
IP			taken care of by 'sendto'
-------------
UDP			taken care of by 'sendto'
-------------
DNS
16	id 		same as question id
1	qr		1 for response
4	opcode	0 for standard query
1	AA		1 for authoritative answer
1	TC		0 for no truncations
1	rd		same as question rd recursion desired
1	ra		0 for no recursion available.
3	z		0 for reserved for future use
4	rcode	0 for no error, 3 for nonexistent host/domain
16	qdc		same as question qdc (hopefully 1)
16	anc		N - number of answer resource records
16	nsc		0 for no ns RRs
16	arc		0 for no additional RRs

-------------
RRs
QUESTION
XX 	qname	same as question qname
16 	qtype	same as question qtype
16 	qclass	same as question qclass
ANSWER * N
XX	name	0xC00C - numchars chars format of name in compressed format
16	type	1 - host address (IP), 12 - domain name
16	class	1 for internet
32	ttl		0 for no caching, 3600 (0xE10) - 1 hour
16	rdlen   L - length of rdata field
XX	rdata	numchars chars format of answer
-------------


CHANGES
Change Log

2002-07-13 @ 19:32 PAcific time - FC
Changed MX record to default precedence to 1 instead of 0 to compensate
for M$ error.

2001-10-29 @ 11:30 - DJH
Added MX capability
added ability to create different servers on different ip addresses,
each as a separate process.

2001-08-20 @ 09:04 -
Added multiple IP address answers.

2001-07-19 @ 13:48 -
Made SDNS version 2.0.  This version provides multiple resource records in the
answer section.  Each request answers with all aliases to the question.  The
first answer RR is the normal answer provided so as to make nslookup happy.
To see all of the aliases, do 'dig <query>', which shows all aliases.

2000-12-01 @ 10:29 -
Made SDNS version 1.0 and started the change log.


convert.pl
#!/usr/bin/perl

# ----------- Default Domain IP Information -----------
# $defaultdomain{".your.domain"} = "1.2.3.4";  <-- Example: Do not use!
require "defaults";
# ----------- End Default Domain IP Information -------                              
$TYPE_IP = 1;
$TYPE_NAME = 12;
$TYPE_MX = 15;
$CLASS_INET = 1;
$COMPRESS = 0xC00C;
$TTL=3600;
$MX_PREF = pack("n",0);
$name_rr = pack("n n n N", $COMPRESS, $TYPE_NAME, $CLASS_INET, $TTL);
$ip_rr = pack("n n n N", $COMPRESS, $TYPE_IP, $CLASS_INET, $TTL);
$mx_rr = pack("n n n N", $COMPRESS, $TYPE_MX, $CLASS_INET, $TTL);

print "/* table.h - Hostname/IP table initialization file for sdns.c */\n\n";
print "#ifndef _SDNS_TABLE_H\n#define _SDNS_TABLE_H\n\n";
print "/* translate sets the string 'answer' equal to the answer of the query,\n";
print "   sets ancount equal to the number of answers in the answer string, and\n";
print "   returns the length of the answer string */ \n";
print "int translate(unsigned short int qtype)\n{\n";
print "
#include <stdio.h>
#include <string.h>
extern char question[];
extern char out[];
extern char answer[];
extern int ancount,arcount ;

#define RECVSIZE 512
#define BUFFSIZE RECVSIZE*4 /* fc - BUFFSIZE better be > 255 bytes bigger than max ethernet frame */
";

for $hostsline (<>) {
	@allnames = ();
	$ancount=2;
	$hostsline =~ y/A-Z/a-z/;
	($ip, $names) = split(/\s+/, $hostsline, 2);
	$ip_rdata = trname($ip,);
	if ((unpack("c", $hostsline) eq "#") || ($ip eq "")) {
		next; }
	else {
		($a, $b, $c, $d) = split(/\./, $ip, 4);
		($car, $cdr) = split(/\s+/, $names, 2);
		if ($car ne "") {
			print "\tif (\t";
			print "!strcmp(question, \"$d.$c.$b.$a.in-addr.arpa\") ||\n";
			print "\t\t!strcmp(question, \"$a.$b.$c.$d\") ||\n\t\t"; }
		while($car ne "") {
			push @allnames,$car;
			print "!strcmp(question, \"$car\") ";
			($car, $cdr) = split(/\s+/, $cdr, 2);
			if ($car ne "") {
				print "||\n\t\t"; }
			else {
				print ") {\n";
				# Generate the answer strings
				$mxname = trmxname($allnames[0]);
				$firstname = trname(shift(@allnames));
				$ip_first_answer = $ip_rr . $ip_rdata . $name_rr .  $firstname;
				$name_first_answer = $name_rr . $firstname . $ip_rr . $ip_rdata;
				$mx_first_answer = $mx_rr . $mxname . $ip_rr . $ip_rdata;
				foreach $hostname (@allnames) {
					$ancount++;
					$firstname = trname($hostname);
					$ip_first_answer .= trrr($hostname) . $firstname;
					$name_first_answer .= trrr($hostname) . $firstname; 
					$mx_first_answer .= trrr($hostname) . $firstname; }
				print "\t\tif (qtype == $TYPE_IP) {\n";
				print "\t\t\tstrncpy(out,\"$allnames[0]\", BUFFSIZE-1);\n";
				print sprintfstr($ip_first_answer);
				print "\n\t\telse if (qtype == $TYPE_MX) {\n";
				print "\t\t\tstrncpy(out,\"$allnames[0]\",BUFFSIZE-1);\n";
				print sprintfstr($mx_first_answer); 
				print "\n\t\telse {\n";
				print "\t\t\tstrncpy(out,\"$ip\",BUFFSIZE-1);\n";
				print sprintfstr($name_first_answer) . " }\n"; } } }
	print "\n"; }

foreach $domain (sort keys %defaultdomain) {
	$ip = $defaultdomain{$domain};
	$defdomain_answer = $ip_rr . trname($ip);
	print "\n/* Default Domain Section */\n";
	print "\t if (strlen(question) > strlen(\"$domain\"))\n";
	print "\t\tif(strncmp(question + (strlen(question) - strlen(\"$domain\")),\n";
	print "\t\t\t\t\"$domain\", strlen(\"$domain\")) == 0) {\n";
	print "\t\t\tstrncpy(out,\"$ip\",BUFFSIZE-1);\n";
	$ancount=1;
	print sprintfstr($defdomain_answer) . "\n"; }

print "\n\treturn 0;\n}\n#endif\n";



sub trname 
{
	my ($domainname) = @_;
	my ($len, $endname, $thing);
	if (($domainname =~ m/(\d+\.\d+\.\d+\.\d+)$/) &&
	   ($domainname =~ m/^(\d+\.\d+\.\d+\.\d+)/)) {
		$len = 4;
		my ($a, $b, $c, $d) = split(/\./, $domainname);
		$endname = pack("n C C C C", $len, $a, $b, $c, $d); }
	else {
		$len = length($domainname); 
		$endname = pack("n", ($len + 2));
		my @splitname = split(/\./,$domainname);
		foreach $thing (@splitname) {
			$len = length($thing);
			$endname .= pack("c", $len) . pack("C$len", unpack("C$len", $thing)); }
		$endname .= pack("c", 0); }
	return $endname;
} 

sub trmxname
{
	my ($domainname) = @_;
	my ($len, $endname, $thing);
	$len = length($domainname) + 4;
	$endname = pack("n n", $len, 0);
	my @splitname = split(/\./,$domainname);
	foreach $thing (@splitname) {
		$len = length($thing);
		$endname .= pack("c", $len) . $thing; }
	$endname .= pack("c", 0);
	return $endname
}

sub trrr
{
	my ($domainname) = @_;
	if (($domainname =~m/(\d+\.\d+\.\d+\.\d+)$/) &&
		($domainname =~ m/^(\d+\.\d+\.\d+\.\d+)/)) {
		return $ip_rr; }
	return $name_rr; 
}

sub sprintfstr
{
	my ($start_string) = @_;
	my ($len, $endstring, $i, $value);
	$len = length($start_string);
	$endstring = "";
	print "\t\t\tsprintf(answer, \"";
	for ($i = 0; $i < $len; $i++) {
		$endstring .= "\%c"; }
	$endstring .= "\"";
	for ($i = 0; $i < $len; $i++) {
		$value = sprintf("%d", unpack("C", substr($start_string, $i, 1)));
		$endstring .= ", $value"; }
	$arcount = $ancount - 1; # one answer, rest will go into additional section		
	$endstring .= ");\n\t\t\tancount = 1;\n\t\t\tarcount = $arcount;\n\t\t\treturn " . $len . "; }";
}


sdns.c
/* RFC 1034 and 1035 server */
#include <stdio.h>
#include <stdlib.h>
#include <errno.h>
#include <sys/types.h>
#include <sys/socket.h>
#include <string.h>   /*for memcpy */
#include <ctype.h>
#include <arpa/nameser.h>  /*for HEADER structure */
#include <time.h>
#include <signal.h>

//#include <sys/socket.h>
//#include <netinet/in.h>
#include <arpa/inet.h>


#define HEADERLENGTH 12
#define RECVSIZE 512
#define BUFFSIZE RECVSIZE*4 /* fc - BUFFSIZE better be > 255 bytes bigger than max ethernet frame */
/* fc - we really need 3*RECVSIZE + 12 (or so) for a message as well */
#define BYTESIZE 255
#define	TIMESTAMPSIZE 64
#define DNSPORT  0x3500
#define DNSlog "logfile"
#define LOG(x, ...) { \
	char message[BUFFSIZE];  \
	FILE *F=fopen(DNSlog,"a+");\
	if (F != NULL) {\
		snprintf(message,BUFFSIZE-1,x, ##__VA_ARGS__);\
		logfile(F,message); \
	} \
	fclose(F); \
}


char  question[BUFFSIZE], answer[BUFFSIZE], out[BUFFSIZE];
int ancount,arcount;


struct sockaddr_in cliaddr;

#include "table.h"
void logfile(FILE * F,char * message) {
	static unsigned int messcount=0,  messcount2=0;	// log queue should be lengthened
	static char oldIP[TIMESTAMPSIZE], oldPort[TIMESTAMPSIZE], oldIP2[TIMESTAMPSIZE], oldPort2[TIMESTAMPSIZE];
	static char oldtimestamp[TIMESTAMPSIZE], oldtimestamp2[TIMESTAMPSIZE], lastmess[BUFFSIZE], lastmess2[BUFFSIZE];

	char timestamp[TIMESTAMPSIZE];
	time_t t=time(NULL);
	strftime(timestamp, 20, "%Y/%m/%d %T", localtime(&t));
	if (!strcmp(message, lastmess)) {
		messcount++;
		if (messcount == 10 || messcount == 50 || messcount == 100 || messcount == 1000 || messcount == 2000 || messcount == 5000)
			fprintf(F, "%s %s:%d sdns %s (%d times)\n",timestamp,inet_ntoa(cliaddr.sin_addr),(cliaddr.sin_port),lastmess,messcount);
		strncpy(oldtimestamp,timestamp,sizeof(oldtimestamp) -1 );}
	else if (!strcmp(message, lastmess2)) {
		messcount2++;
		if (messcount2 == 10 || messcount2 == 50 || messcount2 == 100 || messcount2 == 1000 || messcount2 == 2000 || messcount2 == 5000)
			fprintf(F, "%s %s:%d sdns %s (%d times)\n",timestamp,inet_ntoa(cliaddr.sin_addr),(cliaddr.sin_port),lastmess2,messcount2);
		strncpy(oldtimestamp2,timestamp,sizeof(oldtimestamp) -1);}
	else    {
		if (messcount2 > 1)
			fprintf(F, "%s %s:%s sdns %s (%d times in total)\n",oldtimestamp2,oldIP2,oldPort2,lastmess2,messcount2);
		messcount2=messcount;
		strncpy(lastmess2,lastmess, sizeof(lastmess2)-1) ;
		strncpy(oldIP2, oldIP,sizeof(oldIP2)-1);
		strncpy(oldPort2,oldPort,sizeof(oldPort2)-1);
		strncpy(oldtimestamp2,oldtimestamp,sizeof(oldtimestamp) -1);
		messcount = 1;
		strncpy(lastmess,message,sizeof(lastmess)-1);
		sprintf(oldIP,"%s",inet_ntoa(cliaddr.sin_addr));
		sprintf(oldPort,"%d",(cliaddr.sin_port));
		strncpy(oldtimestamp,timestamp,sizeof(oldtimestamp) -1);
		fprintf(F,"%s %s:%d sdns %s\n",timestamp,inet_ntoa(cliaddr.sin_addr),(cliaddr.sin_port),message);}
}

void dumppacket(char * in_packet, unsigned inpacketsize) {
	FILE *F;
	F=fopen(DNSlog,"a+");
	if (F != NULL) {
		char message[BUFFSIZE];
		snprintf(message,sizeof(message)-1, "Dump: (%d)", inpacketsize);
		unsigned i;
		for (i = 0; i < inpacketsize; i++) {
			unsigned int c=in_packet[i];
			char newmess[8];
			snprintf(newmess,7," %X", c);
			strcat(message,newmess);
		}
		logfile(F, message);
	}			/* This sprintf makes RECVSIZE*3 + 8 the minimum size for BUFFSIZE */
	fclose(F);
}

static void setResponse(HEADER *out_head,int rd, uint rcode, int ancount,int qdcount, int arcount) {
	out_head->rcode = rcode;		/* no such host or domain (we are authoritive) */
	out_head->ancount = ancount;	/* 0 answers (assumed) */
	out_head->qdcount = qdcount;	/* 0 question only (assumed) */
	out_head->nscount = 0x00;	/* 0 number of name server resource records in authority record section */
	out_head->arcount = arcount;	/* 0 number or resource records in the additional records section */
	out_head->qr = 1;		/* Response */
	out_head->opcode = 0;		/* Opcode */
	out_head->aa = 1;		/* Authoritative */
	out_head->tc = 0;		/* Not truncated */
	out_head->rd = rd;	/* Recursion desired is to be copied over (I don't want it...) */
	out_head->ra = 0;		/* No recursion available */
	out_head->unused = 0;		/* Unset unused bits */
}

static int buildListenSocket(int argc, char *argv[]) {
	int sockfd;
	struct sockaddr_in servaddr;
	if ((sockfd = socket(PF_INET, SOCK_DGRAM, 0)) == -1) {
		LOG("Can't create socket");
		fprintf(stderr,"Can't create socket\n");
		return -1;
	}
	if (argc > 1) {
		bzero(&servaddr, sizeof(servaddr));
		servaddr.sin_family = AF_INET;
		servaddr.sin_addr.s_addr = inet_addr(argv[1]);
		servaddr.sin_port = DNSPORT;
	} else {
		bzero(&servaddr, sizeof(servaddr));
		servaddr.sin_family = AF_INET;
		servaddr.sin_addr.s_addr = 0;
		servaddr.sin_port = DNSPORT;
	}
	int retval = bind(sockfd, (struct sockaddr *) &servaddr, sizeof(servaddr));
	if (retval == -1) {
		LOG("Can't Bind");
		fprintf(stderr,"Can't Bind\n");
		return -1;
	}
	return sockfd;
}
static int g_done=0;
static void sigterm_handler(int sig __attribute__((unused)) ) {
	g_done=1;
}

static int setReqTypeRtnIgnore(int qtype,char **reqtypepp){
	int ignore =0;
	char * reqtype;
		switch (qtype) {
			case 1:	reqtype="A";ignore=0;break;
			case 2:	reqtype="NS";ignore=0;break;
			case 3:	reqtype="MD(old)";ignore=NOTIMP;break;
			case 4:	reqtype="MF(old)";ignore=NOTIMP;break;
			case 5:	reqtype="CNAME";ignore=0;break;
			case 6:	reqtype="SOA";ignore=REFUSED;break;
			case 7:	reqtype="MB(old)";ignore=NOTIMP;break;
			case 8:	reqtype="MG(old)";ignore=NOTIMP;break;
			case 9:	reqtype="MR(old)";ignore=NOTIMP;break;
			case 10:reqtype="NULL";ignore=NOTIMP;break;
			case 11:reqtype="WKS";ignore=NOTIMP;break;
			case 12:reqtype="PTR";ignore=0;break;
			case 13:reqtype="HINFO";ignore=REFUSED;break;
			case 14:reqtype="MINFO";ignore=REFUSED;break;
			case 15:reqtype="MX";ignore=0;break;
			case 16:reqtype="TXT";ignore=REFUSED;break;
			case 17:reqtype="RespniblePerson(exper)";ignore=REFUSED;break;
			case 18:reqtype="AndrewFS(exper)";ignore=NOTIMP;break;
			case 19:reqtype="X.25(exper)";ignore=NOTIMP;break;
			case 20:reqtype="ISDN(exper)";ignore=NOTIMP;break;
			case 21:reqtype="RouteThrough(exper)";ignore=NOTIMP;break;
			case 22:reqtype="NSAP(exper)";ignore=NOTIMP;break;
			case 23:reqtype="NSAP-PTR(exper)";ignore=NOTIMP;break;
			case 24:reqtype="SecSig";ignore=REFUSED;break;
			case 25:reqtype="SecKey";ignore=REFUSED;break;
			case 26:reqtype="X.400(exper)";ignore=NOTIMP;break;
			case 27:reqtype="GPS";ignore=REFUSED;break;
			case 28:reqtype="IPV6-P";ignore=-1;break;
			case 29:reqtype="Loc";ignore=REFUSED;break;
			case 30:reqtype="NextDom";ignore=NOTIMP;break;
			case 31:reqtype="Endpoint";ignore=NOTIMP;break;
			case 32:reqtype="Nimrod/SMB-NS";ignore=NOTIMP;break;
			case 33:reqtype="ServerSel/SMB-NodeStat";ignore=NOTIMP;break;
			case 34:reqtype="ATMad";ignore=NOTIMP;break;
			case 35:reqtype="NameAuth";ignore=NOTIMP;break;
			case 36:reqtype="KeyExch";ignore=REFUSED;break;
			case 37:reqtype="CERT";ignore=REFUSED;break;
			case 38:reqtype="IPV6-A";ignore=-1;break;
			case 39:reqtype="Dname";ignore=NOTIMP;break;
			case 40:reqtype="SINK";ignore=NOTIMP;break;
			case 41:reqtype="OPT";ignore=NOTIMP;break;
			case 42:reqtype="APL";ignore=REFUSED;break;
			case 100:reqtype="UINFO";ignore=REFUSED;break;
			case 101:reqtype="UID";ignore=REFUSED;break;
			case 102:reqtype="GID";ignore=REFUSED;break;
			case 103:reqtype="UNSPEC";ignore=REFUSED;break;
			case 249: case -7:reqtype="TransKey";ignore=REFUSED;break;
			case 250: case -6:reqtype="TransSig";ignore=REFUSED;break;
			case 251: case -5:reqtype="IncXfer";ignore=REFUSED;break;
			case 252: case -4:reqtype="AXFR";ignore=REFUSED;break;
			case 253: case -3:reqtype="MAILB";ignore=REFUSED;break;
			case 254: case -2:reqtype="MAILA";ignore=REFUSED;break;
			case 255: case -1:reqtype="ALL";ignore=0;break;
			default:reqtype="Unknown";ignore=NOTIMP;break;
		}
		(*reqtypepp) = reqtype;
		return ignore;
}

static void listenAndRespondLoop(int sockfd) {
	char in_packet[BUFFSIZE], out_packet[BUFFSIZE];
	while (!g_done) {
		int i;
		for (i = 0; i < BUFFSIZE; i++) {
			out_packet[i] = '\0';
			in_packet[i] = '\0';
			question[i] = '\0';
			answer[i] = '\0';
		}
		socklen_t len = sizeof(cliaddr);
		int size=recvfrom(sockfd, in_packet, RECVSIZE, 0, (struct sockaddr *) &cliaddr, (socklen_t *)&len);
		//unsigned int inpacketsize=size;
		if (size > RECVSIZE) {
		//	inpacketsize=RECVSIZE;
			LOG("System call screwed up - packet too big");
			continue;
		}
		if (size <= HEADERLENGTH) {
			LOG("packet too small");
			continue;
		}
		HEADER *in_head = (HEADER*) &in_packet;
		HEADER *out_head = (HEADER*) &out_packet;
		if (in_head->opcode != QUERY) {
			LOG("Not a query (opcode != 0)");
			continue;
		}
		if (in_head->qr != 0) {
			LOG("Not a query (qr != 0)");
			continue;
		}
		if (in_head->qdcount == 0) {
			LOG("Question count == 0");
			continue;
		}
		unsigned int qindex = 0;
		size = 0;
		char *read_offset = in_packet + HEADERLENGTH;
		while(read_offset[size] != '\0'){		/* pre-zero'ed packet will run into a zero after end of packet if nowhere else */
			unsigned char charstilldot = (unsigned char)read_offset[size++];	/* the size so far... */
			if ((qindex != 0) && (charstilldot != 0))
				question[qindex++] = '.';
			while (charstilldot--) {
				question[qindex++] = tolower(read_offset[size++]);
			}
		}
		size++;	/* there could be real hazardous characters in the question... */
		if (size <= 1) {
			LOG("Illegal name size (too small)");
			continue;
		}
		if (size > 255) {
			LOG("Illegal name size (too big)");
			continue;
		}
		unsigned int port=cliaddr.sin_port;
		if (port == NAMESERVER_PORT) {
			LOG("Port 53 as source");
			continue;
		} /* does zone transfer use port 53 source?*/
		int qtype = read_offset[size] * 256 + read_offset[size + 1]; /* Qtype */
		out_head->id = in_head->id;
		int ans_size = translate(qtype);
		int ignore = 0;
		char *reqtype="NOT_SET";
		ignore = setReqTypeRtnIgnore(qtype,&reqtype);
		if (-1 == ignore) {
			LOG("Ignoring %s (%s:%d) => %s", question, reqtype, qtype, out);
			continue;
		}
		if (ans_size == 0) {
			LOG("lookup %s (%s:%d) - not in table", question, reqtype, qtype);
			setResponse(out_head,in_head->rd, 3, 0,0,0);
			size=-4;
		}			/* no question, no answer, no problem */
		else if (ignore != 0) {
			LOG("Code %s (%s:%d) => %d", question, reqtype, qtype, ignore);
			setResponse(out_head,in_head->rd, ignore, 0,0,0);
			size=-4;
		}
		else	{
			LOG("lookup %s (%s:%d) => %s", question, reqtype, qtype, out);
			setResponse(out_head,in_head->rd, 0, htons(ancount),0x100,htons(arcount));
			memcpy(out_packet + HEADERLENGTH, in_packet + HEADERLENGTH, size + 4); /* Copy question section into out packet */
			read_offset = out_packet + size + HEADERLENGTH + 4; out_head->rcode = 0; /* Point read_offset to answer section */
			memcpy(read_offset, answer, ans_size);
		}     /* Copy answer section into out packet */

		int retval = sendto(sockfd,&out_packet,(HEADERLENGTH + size + 4 + ans_size), 0, (const struct sockaddr *) &cliaddr, len);
		if (retval == -1)
			LOG("could not send");
	}
}

int main(int argc __attribute__((unused)), char *argv[] __attribute__((unused))) {
	signal (SIGTERM, sigterm_handler);
	signal (SIGINT, sigterm_handler);
	LOG("Starting sdns");
	int sockfd = buildListenSocket(argc,argv);
	if (-1 == sockfd)
		exit(EXIT_FAILURE);
	listenAndRespondLoop(sockfd);
	exit(EXIT_SUCCESS);
}
